File: C++/23_elements_occuring_more_than_k_times.cpp
/*
Question: find elements occurring more than n/k times, n=length, and k-constant
Idea:
Counting occurrences of each element, and checking if any surpass the threshold
*/
#include<bits/stdc++.h>
using namespace std;


vector<int> elementsAppearance(int arr[], int n, int k)
{
    map<int, int> mp;
    vector<int> ans;

    //Counting each element's occurrences
    for (int i = 0;i < n;++i)
        mp[arr[i]]++;

    for (auto it : mp)
    {
        //if count reaches threshold , append in result
        if (it.second > (n / k))
            ans.push_back(it.first);
    }
    return ans;
}

int main()
{
    int arr[] = { 1, 1, 2, 2, 3, 5, 4, 2, 2, 3, 1, 1, 1 };
    int n = sizeof(arr) / sizeof(arr[0]);
    int k = 4;
    vector<int>ans = elementsAppearance(arr, n, k);
    for (int i = 0;i < ans.size();++i)
        cout << ans[i] << " ";
    cout << "\n";
}

File: Python/Array/23_elements_occuring_more_than_k_times.py
"""
Question: find elements occurring more than n/k times, n=length, and k-constant
Idea:
Counting occurrences of each element, and checking if any surpass the threshold

"""
from collections import Counter


def elementsAppearance_using_inbuilt_function(arr, n, k):
    """
    Using python outing inbuilt function
    """
    x = n // k

    # Counting frequency of every element using Counter
    mp = Counter(arr)

    # Traverse the map and print all the elements with occurrence more than n/k times
    for it in mp:
        if mp[it] > x:
            print(it)


def elementsAppearance(arr, n, k):
    frequency = {}
    ans = []
    for i in range(n):
        # Counting each element's occurrences
        if arr[i] in frequency:
            frequency[arr[i]] += 1
        else:
            frequency[arr[i]] = 1

    # if count reaches threshold , append in result
    for i in frequency:
        if frequency[i] > n / k:
            ans.append(i)

    return ans


if __name__ == '__main__':
    a = [1, 1, 2, 2, 3, 5, 4, 2, 2, 3, 1, 1, 1]
    length = len(a)
    key = 4
    ans = (elementsAppearance(a, length, key))
    print(ans)
